### PR TITLE
resolve redirects bulk

### LINKED ISSUE #6511: Resolve Redirects Bulk
https://docs.google.com/document/d/1d9DxkqhPP_Rs01ANedMeWjrTgd7LEiDrWb-WNtA9fPg/edit

### Describe the problem that you'd like solved

Manual Bulk process: update all redirect references in db since X, older than 1 week 
We have a script in scripts/ called update_redirect_references.py which imports the core models from P1.
Script creates or updates a statefile.pid which references the last cutoff timestamp (everything before this date has already been processed)
script loops over pages of redirects from https://openlibrary.org/query.json?type=/type/redirect&location&limit=1000&last_modified&sort=-last_modified to fetch ~unprocessed redirects (i.e. after the time cutoff + ignoring any redirects that have been created less than 1 week ago)
Results of API: { type:/type/redirect, key: /works/OL123W, location: /works/OL42W }
Of the resulting redirects (older than 1 week) we will then call the update_redirect_references method on each core model on each redirect ID
Run over entire redirect history

### PR DESCRIPTION
**Pull request by mekarpeles**
Closes #6511 

### Testing

Run with:
```
import web; import infogami; from openlibrary.config import load_config; load_config('/olsystem/etc/openlibrary.yml');infogami._setup();from infogami import config;from openlibrary.core.models import Work

Work.resolve_redirects_bulk(batch_size=1000, max_limit=10000, test=True)
```

### CODE REVIEW

File: openlibrary/core/models.py [hunk 1]
```diff
@@ -593,6 +596,103 @@ def get_redirect_chain(work_key: str) -> list:
...
+                olid, new_olid, _test=test
+            )
+        return summary
+
+    @classmethod
+    def resolve_redirects_bulk(
+            cls,
+            batch_size=1000,
+            start_offset=0,
+            pages=None,
```
Comment (cdrini) [inline openlibrary/core/models.py:650]:
```suggestion
            pages=1,
```

File: openlibrary/core/models.py [hunk 2]
```diff
@@ -593,6 +596,103 @@ def get_redirect_chain(work_key: str) -> list:
...
+                page+1,
+                pages or "∞",
+                offset
+            )
+            work_redirect_ids = web.ctx.site.things({
+                "type": "/type/redirect",
+                "key~": "/works/*",
+                "limit": batch_size,
+                "offset": offset
+            })
```
Comment (cdrini) [inline openlibrary/core/models.py:677]: We can sort this in reverse order of last_modified, which should simplify stuff. This works in the URL, and might be `sort: '-last_modified'` in the python?

http://openlibrary.org/query.json?type=/type/redirect&key~=/works/*&sort=-last_modified&*=

File: openlibrary/core/models.py [hunk 3]
```diff
@@ -593,6 +596,103 @@ def get_redirect_chain(work_key: str) -> list:
...
+        batch_size - how many records to fetch per batch
+        start_offset - what offset to start from
+        pages - how many pages of batch_size to process
+        end_date - ignore redirects created after this date
+        start_date - ignore redirects created before this date
+        test - don't resolve stale redirects, just identify them
+        """
+        page = 0
+        offset = start_offset
+        while not pages or page < pages:
```
[the remark this answers is not in the document]
Comment (cdrini) [reply openlibrary/core/models.py:665]: Alternatively you could do:

```py
for offset in range(start_offset, max_limit, batch_size):
```

eg:

```py
>>> list(range(0, 100, 10))
[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]
```

This let's us remove the `page` abstraction which is one less index to worry about. But whichever is better :+1:

File: openlibrary/core/models.py [hunk 4]
```diff
@@ -593,6 +596,103 @@ def get_redirect_chain(work_key: str) -> list:
...
+                "[resolving-redirects] batch:%i of %i, offset:%i",
+                page+1,
+                pages or "∞",
+                offset
+            )
+            work_redirect_ids = web.ctx.site.things({
+                "type": "/type/redirect",
+                "key~": "/works/*",
+                "limit": batch_size,
+                "offset": offset
```
Comment (mekarpeles) [inline openlibrary/core/models.py:675]:
```suggestion
                "offset": offset,
                "sort": "-last_modified"
```

**Comment by mekarpeles** [comment]
@cclauss do you mind taking a look at https://results.pre-commit.ci/run/github/69609/1656295879.B53XpkRZQJSnqaoW_nS5KQ and seeing what type of changes need to be made for mypy? This would be a big help, we're trying to wrap this up for H1

File: openlibrary/core/models.py [hunk 5]
```diff
@@ -593,6 +596,108 @@ def get_redirect_chain(work_key: str) -> list:
                 resolved_key = thing.key
         return redirect_chain
 
+    @classmethod
+    def resolve_redirect_chain(cls, work_key: str, test: bool=False):
+        summary = {
```
Comment (cclauss) [inline openlibrary/core/models.py:601]:
```suggestion
    def resolve_redirect_chain(cls, work_key: str, test: bool = False):
        summary: dict[str, Any] = {
```

File: openlibrary/core/models.py [hunk 6]
```diff
@@ -593,6 +596,108 @@ def get_redirect_chain(work_key: str) -> list:
...
+            {
+                "key": thing.key,
+                "occurrences": {},
+                "updates": {}
+            } for thing in redirect_chain
+        ]
+        summary['resolved_key'] = redirect_chain[-1].key
+
+        for r in summary['redirect_chain']:
+            olid = r['key'].split('/')[-1][2:-1]
```
Comment (cclauss) [inline openlibrary/core/models.py:617]: OPTIONAL: Explain why  `[2:-1]`
```suggestion
            olid = r['key'].split('/')[-1][2:-1]  # 'OL1234x' --> '1234'
```

File: openlibrary/core/models.py [hunk 7]
```diff
@@ -593,6 +596,108 @@ def get_redirect_chain(work_key: str) -> list:
...
+
+    @classmethod
+    def resolve_redirects_bulk(
+            cls,
+            batch_size=1000,
+            start_offset=0,
+            max_limit=1000,
+            grace_period_days=7,
+            cutoff_date=datetime.datetime(year=2017, month=1, day=1),
+            test=True
```
Comment (cclauss) [inline openlibrary/core/models.py:653]: Optional: naming...
```suggestion
            dry_run=True
```

**Comment by jimman2003** [comment]
The any type needs an import: [from typing import Any](https://docs.python.org/3/library/typing.html#the-any-type)

**Review by cclauss — APPROVED** [review]

File: openlibrary/core/models.py [hunk 8]
```diff
@@ -593,6 +597,108 @@ def get_redirect_chain(work_key: str) -> list:
...
+    ):
+        """
+        batch_size - how many records to fetch per batch
+        start_offset - what offset to start from
+        pages - how many pages of batch_size to process
```
Comment (cdrini) [inline openlibrary/core/models.py:659]: Update docs

File: openlibrary/core/models.py [hunk 9]
```diff
@@ -593,6 +597,108 @@ def get_redirect_chain(work_key: str) -> list:
...
+        grace_period_days - ignore redirects created within period of days
+        cutoff_date - ignore redirects created before this date
+        test - don't resolve stale redirects, just identify them
+        """
+        pos = 0
+        grace_date = (
```
Comment (cdrini) [inline openlibrary/core/models.py:665]: This might be clearer as `max_date`, in that we've basically just got a date range to process.

**Review by cdrini — COMMENTED** [review]
The logic looks good to me awaiting testing!

File: openlibrary/plugins/admin/code.py [hunk 10]
```diff
@@ -220,49 +220,12 @@ def main(self, test=False):
         if test is False and params.test:
```
Comment (cclauss) [inline openlibrary/plugins/admin/code.py:220]: https://github.com/internetarchive/openlibrary/blob/master/pyproject.toml#L20

**Comment by mekarpeles** [comment]
Test:

````
Work.resolve_redirects_bulk(batch_size=10, start_offset=50000, limit=2, test=True)
```

```
2022-07-15 20:03:33 [387] [openlibrary.core] [INFO] [update-redirects] Batch 1: #50000 of 50002
0.0 (1): SELECT * FROM thing WHERE key='/type/redirect'
0.0 (1): SELECT * FROM thing WHERE key='/type/redirect'
0.0 (1): SELECT set_config('statement_timeout', '60000', false)
2.27 (2): SELECT thing.key FROM thing WHERE thing.type = 17729060 AND thing.key LIKE '/works/%' ORDER BY thing.last_modified desc LIMIT 10 OFFSET 50000
2022-07-15 20:03:37 [387] [openlibrary.core] [INFO] [update-redirects] Update: #50001 </works/OL14911443W> {'key': '/works/OL14911443W', 'redirect_chain': [{'key': '/works/OL14911443W', 'occurrences': {'readinglog': 0, 'ratings': 0, 'booknotes': 0, 'observations': 0}, 'updates': {'readinglog': {'rows_changed': 0, 'rows_deleted': 0, 'failed_deletes': 0}, 'ratings': {'rows_changed': 0, 'rows_deleted': 0, 'failed_deletes': 0}, 'booknotes': {'rows_changed': 0, 'rows_deleted': 0, 'failed_deletes': 0}, 'observations': {'rows_changed': 0, 'rows_deleted': 0, 'failed_deletes': 0}}}, {'key': '/works/OL14849978W', 'occurrences': {'readinglog': 1, 'ratings': 0, 'booknotes': 0, 'observations': 0}, 'updates': {'readinglog': {'rows_changed': 1, 'rows_deleted': 0, 'failed_deletes': 0}, 'ratings': {'rows_changed': 0, 'rows_deleted': 0, 'failed_deletes': 0}, 'booknotes': {'rows_changed': 0, 'rows_deleted': 0, 'failed_deletes': 0}, 'observations': {'rows_changed': 0, 'rows_deleted': 0, 'failed_deletes': 0}}}], 'resolved_key': '/works/OL14849978W'}
2022-07-15 20:03:37 [387] [openlibrary.core] [INFO] [update-redirects] Done: #50002 of 50002
```

**Comment by mekarpeles** [comment]
For next week -- I just ran a few tests of the resolve redirects bulk code. Nothing should change in terms of the original algorithm which previously merged. All core feedback was addressed from @cclauss + @cdrini. I ran a single case successfully and several different tests with the `test` flag.

I'd like to run a few more tests w/ staff next week before we put this into practice. Also, to discuss how this should be run (e.g. 1-off?)

**Pull request merged**